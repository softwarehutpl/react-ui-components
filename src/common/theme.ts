const theme: { [key: string]: { [key: string]: { [key: string]: string } } } = {
  colors: {
    primary: {
      base: '#000000',
      light: '#ffffff',
      dark: '#a8b3bd',
    },
    secondary: {
      base: '#3690e3',
      light: '#75b0e6',
      dark: '#034d91',
    },
    error: {
      base: '#f44336',
      light: '#f56c62',
      dark: '#a1140a',
    },
    warning: {
      base: '#faa72d',
      light: '#ffc36b',
      dark: '#995b00',
    },
    info: {
      base: '#7979d4',
      light: '#9797d1',
      dark: '#4c4cc7',
    },
    success: {
      base: '#4caf50',
      light: '#88c28a',
      dark: '#124f14',
    },
    disabled: {
      base: '#d9d9d9',
      light: '#f0eded',
      dark: '#8c8989',
    },
  },
  // TODO: change it in dedicated task for that (change ease-out for example and add more options)
  transitions: {
    all: {
      fast: 'all 0.5s ease-out',
      mid: 'all 2s ease-out',
      slow: 'all 4s ease-out',
    },
    background: {
      fast: 'background-color 0.5s ease-out',
      mid: 'background-color 2s ease-out',
      slow: 'background-color 4s ease-out',
    },
    fontColor: {
      fast: 'color 0.5s ease-out',
      mid: 'color 2s ease-out',
      slow: 'color 4s ease-out',
    }
  }
};

export default theme;
