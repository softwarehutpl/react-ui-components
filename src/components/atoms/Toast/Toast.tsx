import React, { useEffect } from 'react';
import styled from 'styled-components';
import {COLOR_BLACK, COLOR_WHITE} from '../../../constants/colors';

interface IToastProps {
  message: string;
  title?: string;
  className?: string;
  onClose: () => void;
  duration?: number;
  onClick?: () => void;
}

interface IToastStyleProps {
  width?: number;
  color?: string;
  backgroundColor?: string;
  fontColor?: string;
  position?: 'top' | 'bottom';
  padding?: number;
}

const Toast = ({ message, title, className, duration, onClose, onClick }: IToastProps) => {
  useEffect(() => {
    const timeout = setTimeout(onClose, duration);
    return () => {
      clearTimeout(timeout);
    };
  }, [duration, onClose]);

  return (
    <div className={className} onClick={onClick}>
      <h4>{title}</h4>
      <p>{message}</p>
    </div>
  );
};

const defaultProps = {
  width: 300,
  color: 'primary',
  position: 'bottom',
  padding: 15,
  duration: 3000,
};

Toast.defaultProps = defaultProps;

export default styled(Toast)<IToastStyleProps>`
  ${({
    theme,
    width = defaultProps.width,
    color = defaultProps.color,
    backgroundColor,
    fontColor,
    position = defaultProps.position,
    padding = defaultProps.padding,
  }) => ({
    width: `${width}px`,
    'background-color': backgroundColor || COLOR_WHITE,
    color: fontColor || COLOR_BLACK,
    position: 'fixed',
    bottom: position === 'bottom' ? '50px' : 'auto',
    top: position === 'top' ? '50px' : 'auto',
    left: `calc(50vw - ${width / 2}px)`,
    padding: `${padding}px`,
    'border-radius': '5px',
    'border-top': `7px solid ${theme.colors[color].base}`,
    'box-shadow': '0px 4px 10px 4px rgba(217,217,217,0.6)',
    h4: {
      'margin-top': 0,
      'margin-bottom': '0.5em',
      'text-align': 'left',
    },
    p: {
      'font-size': '0.8em',
      'line-height': '1.5em',
      margin: 0,
      'text-align': 'justify',
    },
  })}
`;
