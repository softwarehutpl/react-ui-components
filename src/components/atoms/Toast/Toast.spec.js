import React from 'react';
import { create, act } from 'react-test-renderer';
import 'jest-styled-components';
import theme from '../../../common/theme';
import Toast from './Toast';

describe('Toast component', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  it('should render correctly', () => {
    const tree = create(<Toast theme={theme} message="Lorem ipsum" />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render toast with given title and text', () => {
    const component = create(<Toast message="Lorem ipsum" title="title" theme={theme} />);
    const instance = component.root;
    const header = instance.findByType('h4');
    const message = instance.findByType('p');
    expect(header.props.children).toBe('title');
    expect(message.props.children).toBe('Lorem ipsum');
  });

  it('should render toast with default style', () => {
    const component = create(<Toast message="Lorem ipsum" title="title" theme={theme} />);
    const toast = component.toJSON();
    expect(toast).toHaveStyleRule('width', '300px');
    expect(toast).toHaveStyleRule('background-color', '#fff');
    expect(toast).toHaveStyleRule('color', '#000');
    expect(toast).toHaveStyleRule('padding', '15px');
  });

  it('should render toast with custom style', () => {
    const component = create(
      <Toast
        message="Lorem ipsum"
        title="title"
        width={200}
        backgroundColor="red"
        fontColor="yellow"
        theme={theme}
        padding={20}
      />
    );
    const toast = component.toJSON();
    expect(toast).toHaveStyleRule('width', '200px');
    expect(toast).toHaveStyleRule('background-color', 'red');
    expect(toast).toHaveStyleRule('color', 'yellow');
    expect(toast).toHaveStyleRule('padding', '20px');
  });

  it('should render toast in default bottom position', () => {
    const component = create(<Toast theme={theme} />);
    const toast = component.toJSON();
    expect(toast).toHaveStyleRule('bottom', '50px');
  });

  it('should render toast in top position', () => {
    const component = create(<Toast theme={theme} position="top" />);
    const toast = component.toJSON();
    expect(toast).toHaveStyleRule('top', '50px');
    expect(toast).toHaveStyleRule('bottom', 'auto');
  });

  // TODO: fix this test - search more about jest.advanceTimersByTime and jest.runAllTimers and
  // error: 'Cannot read property 'apply' of undefined'
  it('should call onClose method with default delay', () => {
    const mockOnClose = jest.fn();
    const component = create(<Toast theme={theme} onClose={mockOnClose} />);
    expect(mockOnClose).not.toBeCalled();
    jest.advanceTimersByTime(3000);
    expect(mockOnClose).toBeCalled();
  });

  it('should call onClick method', () => {
    const mockOnClick = jest.fn();
    const component = create(<Toast onClick={mockOnClick} theme={theme} />);
    const instance = component.root;
    const toast = instance.findByType(Toast);
    act(() => {
      toast.props.onClick();
    });
    expect(mockOnClick.mock.calls.length).toEqual(1);
  });
});
