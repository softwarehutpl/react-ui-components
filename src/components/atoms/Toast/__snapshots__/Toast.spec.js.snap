// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast component should render correctly 1`] = `
.c0 {
  width: 300px;
  background-color: #fff;
  color: #000;
  position: fixed;
  bottom: 50px;
  top: auto;
  left: calc(50vw - 150px);
  padding: 15px;
  border-radius: 5px;
  border-top: 7px solid #000000;
  box-shadow: 0px 4px 10px 4px rgba(217,217,217,0.6);
}

.c0 h4 {
  margin-top: 0;
  margin-bottom: 0.5em;
  text-align: left;
}

.c0 p {
  font-size: 0.8em;
  line-height: 1.5em;
  margin: 0;
  text-align: justify;
}

<div
  className="c0"
>
  <h4 />
  <p>
    Lorem ipsum
  </p>
</div>
`;
